### PR TITLE
Initial can-reflect integration for can-define

### PR DESCRIPTION
**Pull request by bmomberger-bitovi**
* can-define.js: add `@@can.onKeyValue` and `@@can.offKeyValue` symbols to defined objects (point to same functions as `on`/`bind`/`addEventListener` and `off`/`unbind`/`removeEventListener` respectively)
* define-helpers.js: make `_instanceDefinitions` a non-enumerable property in all cases, which avoids some test breakages
* map.js: use `can-reflect.eachKey` to iterate properties during `setProps`
* map.js: replace "set" and "attr" and "replace" membership tests with `can-reflect.isMapLike` and `can-reflect.isListLike`, similarly replace non-map-like object test with `can-reflect.isValueLike`
* map.js: replace property merges and sets with `can-reflect.setValue` and `can-reflect.setKeyValue
* map.js: add registered symbols to DefineMap prototype: "@@can.getKeyValue", "@@can.setKeyValue", "@@can.getValue", "@@can.setValue", "@@can.isMapLike", "@@can.isListLike", and "@@can.isValueLike"
* map.js: add iterator from can-define.js as the iterator for DefineMaps using `canSymbol.iterator`
* list.js: in set-multiple case in `set`, use `can-reflect.isListLike` to test arguments, and `can-reflect.setValue`/`can-reflect.eachKey`/`can-reflect.setKeyValue` to set values
* list.js: in `concat` use `can-reflect.isListLike` to test arguments for arrayness, and `can-reflect.eachIndex` to iterate over each argument.
* list.js: in `forEach`, `filter`, and `map`, use `can-reflect.eachIndex` to iterate the list and `can-reflect.call` to call the callback
* list.js: use can-reflect getters and setters in deprecated `attr`
* list.js: add registered symbols to DefineMap prototype: "@@can.getKeyValue", "@@can.setKeyValue", "@@can.getValue", "@@can.setValue", "@@can.isMapLike", "@@can.isListLike", and "@@can.isValueLike"
* list.js: add a simple iterator to DefineLists, using `canSymbol.iterator`

UPDATE:

* can-define.js:  use Observation instead of Compute for properties with getters, unless the getter is async (i.e. declares 2 or more params).

### CODE REVIEW

File: can-define.js [hunk 1]
```diff
@@ -52,6 +53,25 @@ var eachPropertyDescriptor = function(map, cb){
 };
 
 
+function trapSets(observableValue) {
```
Comment (justinbmeyer) [inline can-define.js]: please add some inline docs on what this is doing, it's purpose.

File: can-define.js [hunk 2]
```diff
@@ -668,7 +710,9 @@ assign(eventsProto, {
 		if (computedBinding && computedBinding.compute) {
 			if (!computedBinding.count) {
 				computedBinding.count = 1;
-				computedBinding.compute.addEventListener("change", computedBinding.handler);
+				// TODO use canReflect.onValue when can-compute supports reflection
```
Comment (justinbmeyer) [inline can-define.js]: Can this comment be removed?

File: define-helpers/define-helpers.js [hunk 3]
```diff
@@ -41,7 +41,12 @@ var defineHelpers = {
 		// next if it's already on this instances
 		var instanceDefines = map._instanceDefinitions;
 		if(!instanceDefines) {
-			instanceDefines = map._instanceDefinitions = {};
+			Object.defineProperty(map, "_instanceDefinitions", {
```
Comment (justinbmeyer) [inline define-helpers/define-helpers.js]: why is this needed?

File: list/list.js [hunk 4]
```diff
@@ -261,15 +262,15 @@ var DefineList = Construct.extend("DefineList",
 			}
 			// otherwise we are setting multiple
 			else {
-				if (isArray(prop)) {
+				if (canReflect.isListLike(prop)) {
 					if (value) {
-						this.replace(prop);
+						canReflect.setValue(this, prop);
```
Comment (justinbmeyer) [inline list/list.js]: nice

Comment (justinbmeyer) [inline list/list.js]: though I wonder about `.splice` ... we might need something for that in can-reflect

File: list/list.js [hunk 5]
```diff
@@ -857,8 +858,8 @@ assign(DefineList.prototype, {
 	forEach: function(cb, thisarg) {
 		var item;
 		for (var i = 0, len = this.length; i < len; i++) {
-			item = this.get(i);
-			if (cb.call(thisarg || item, item, i, this) === false) {
+			item = canReflect.getKeyValue(this, i);
```
Comment (justinbmeyer) [inline list/list.js]: I think `.get()` should still be used.  It's ok to call methods on ourselves that we know must exist (faster this way too).

File: list/list.js [hunk 6]
```diff
@@ -1004,8 +1005,8 @@ assign(DefineList.prototype, {
 	map: function(callback, thisArg) {
 		var mappedList = [],
 			self = this;
-		this.each(function(item, index, list) {
-			var mapped = callback.call(thisArg | self, item, index, self);
+		canReflect.eachIndex(this, function(item, index, list) {
+			var mapped = canReflect.call(callback, thisArg | self, item, index, self);
```
Comment (justinbmeyer) [inline list/list.js]: I think we might be able to get away with `callback.call`.  I have to think about if we need to support arbitrary function like things here and now.

Comment (bmomberger-bitovi) [inline list/list.js]: I guess the question on my mind was was how `apply`/`call` got added to can-reflect if we didn't really need them yet for non-natively-callable function-likes

File: define-helpers/define-helpers.js (hunk 3 above)
Comment (bmomberger-bitovi) [inline define-helpers/define-helpers.js]: It shouldn't be writable.

File: list/list.js (hunk 6 above)
Comment (justinbmeyer) [inline list/list.js]: The only thing I had in mind was things like a call expression. But it might not actually be useful because it represents a value, not a method that can be called.

Comment (bmomberger-bitovi) [inline list/list.js]: Got it.  I think it's better to undo this for now and do `callback.call`

File: define-helpers/define-helpers.js (hunk 3 above)
Comment (bmomberger-bitovi) [inline define-helpers/define-helpers.js]: I don't remember the details now, but some stuff was getting screwed up because `_instanceDefinitions` was included in `.forEach()` loops.
